Refactor onnx

Vimalakirti opened the pull request:
This PR refactors `onnx.rs`, adding the following features:
- Improved readability
- Ignored optional input fields of ONNX layers
- Computed precomputable constants without padding (which might cause problems for layers like `Concat`)
- Returned `fake_input` to facilitate testing the Graph (this tool spares us the effort of creating inputs for various ONNX files)

Vimalakirti committed:
refactor

ddkang commented on src/main.rs:
```diff
@@ -109,11 +109,8 @@ fn verify(srs: &SRS, graph: &Graph) {
-  let input: Vec<Fr> = (0..4).map(|_| Fr::from(rng.gen_range(-4..4))).collect();
-  let input = ArrayD::from_shape_vec(vec![1, 4], input).unwrap();
-  let inputs = vec![&input];
+  let (mut graph, models, fake_inputs) = onnx::load_file("sample.onnx");
```
What exactly is fake_inputs and what is the contract?

Vimalakirti replied:
`fake_inputs` are not real inputs read from a dataset. Instead, they are tensors filled with random elements, whose shapes match those described in the input tensors of an ONNX model. Generating these when loading an ONNX file saves us from creating different input tensors ourselves. About the contract, they should have the same shape and data type as the ONNX model's input tensors and should be consistently usable for testing and validation purposes within our system.
I'll update the script and add comments to make it clear.

liliatangxy replied:
I don't think we should include input generation as part of the core onnx file loading. Instead, we can separate this out into a utility function or function that is just used for testing purposes to do the same.

We might want to determine a way to make it easier for a developer to add their own inputs without needing to guess the input order.

ddkang commented on src/onnx.rs:
```diff
@@ -81,108 +86,240 @@ pub fn load_file(filename: &str) -> (Graph, Vec<ArrayD<Fr>>) {
-      });
-    }
+// This function is used for generating fake inputs for onnx models
+// It is only for testing purposes
```
See above about the contracts

Vimalakirti committed:
fix indexing bugs and rename variables

Vimalakirti commented:
The new commit adds the following features:
- fixed indexing bugs in `graph.rs` when feeding `inputs` for `encodeOutputs`, `prove`, and `verify`
- added comments for `fake_inputs`
- renamed confusing variable names

ddkang commented:
@iantimmis we're beginning to refactor the ONNX parsing. It would be great if you could check to see if this improves readability!

liliatangxy commented on src/onnx.rs:
```diff
@@ -81,108 +87,252 @@ pub fn load_file(filename: &str) -> (Graph, Vec<ArrayD<Fr>>) {
+  .unwrap()
+}
+
+// This function is used for updating the graph of onnx after creating the local graph
```
Can you add more comments so that we know what the intended logic of the operation as well as what the arguments are referring to?

liliatangxy commented on src/onnx.rs:
```diff
@@ -81,108 +87,252 @@ pub fn load_file(filename: &str) -> (Graph, Vec<ArrayD<Fr>>) {
+}
+
+// This function is used for processing a layer of onnx models
+fn process_node(
```
More comments for this fn too would be nice

Vimalakirti committed:
addressed comments

Vimalakirti committed:
cleanup
